File: 28-implement-strstr/28-implement-strstr.cpp
class Solution {
public:
    int strStr(string haystack, string needle) {
        int n = needle.size();
        if (n == 0)
            return 0;
        if (haystack.size() < n)
            return -1;
        
        for (int i = 0; i <= haystack.size() - n; i++) {
            if (haystack[i] == needle[0] && haystack[i + n - 1] == needle[n - 1]) {
                for (int j = 0; j < n && haystack[i + j] == needle[j]; j++)
                    if (j == n - 1)
                        return i;
            }
        }
        return -1;
    }
};

File: 28-implement-strstr/28-implement-strstr.java
class Solution {
    public int strStr(String haystack, String needle) {
        int n = needle.length();
        if(needle.isEmpty()) 
            return 0;
       
       for(int i = 0; i <= haystack.length() - needle.length(); i++){

            if(haystack.charAt(i) == needle.charAt(0) && haystack.charAt(i + n - 1) == needle.charAt(n - 1)) { //extra validation

                for(int j = 0; j < needle.length() && haystack.charAt(i +j) == needle.charAt(j); j++) {

                    if(j == needle.length() -1)
                        return i;
                    
                }
            }
       }

       return -1;
    }
}